### PR TITLE
Add Paypal payment gateway

### PR DESCRIPTION
**Pull request by albertojacini**
I want to merge this change because I would love to be able to receive Paypal payments. Braintree has refused our application because we are a marketplace.

Follows #3469

I wrote a very basic plugin for the integration of Paypal. It only supports one shot authorize/capture and refund. I hope it can help. I can work further on it if needed.

Thank sooooooooo much to the team for creating and sharing Saleor!

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by maarcingebala** [comment]
Hey @albertojacini, sorry for no response from us here, the team is constantly busy with new features from the roadmap and various bugfixes, and sometimes it's really hard to also handle PRs from the community for us. The code looks very solid at a first glance and I'll try to test it locally first with some Paypal sandbox account. I should have some feedback and decision whether we want to merge PayPal support to master next week ✌️.

**Comment by albertojacini** [comment]
Thanks a lot @maarcingebala, that would be great :+1: !!.

**Comment by Sarowar-neo** [comment]
I'm waiting for paypal integration ...... hurry up guys ! Thanks in advance.

**Comment by cyberdemic** [comment]
Nice job! Would love to see this also

**Comment by f0rmatme** [comment]
Thanks for this!

File: saleor/payment/gateways/paypal/__init__.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        return GatewayResponse(
+            is_success=True,
+            action_required=False,
+            kind=TransactionKind.CAPTURE,
+            amount=decimal.Decimal(
+                response.result.purchase_units[0].payments.captures[0].amount.value
+            ),
+            currency=response.result.purchase_units[0]
+            .payments.captures[0]
+            .amount.currency_code,
```
Comment (maarcingebala) [inline saleor/payment/gateways/paypal/__init__.py:88]: Here we could have `currency=transaction.amount.value`.

File: saleor/payment/gateways/paypal/__init__.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            transaction_id=transaction.id,
+            error=None,
+            # raw_response=response,
```
Comment (maarcingebala) [inline saleor/payment/gateways/paypal/__init__.py:91]: If this is not needed, I would remove the commented line.

File: saleor/payment/gateways/paypal/__init__.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+def refund(payment_information: PaymentData, config: GatewayConfig) -> GatewayResponse:
+    client = get_paypal_client(**config.connection_params)
+    request = CapturesRefundRequest(payment_information.token)
+    try:
+        response = client.execute(request)
+    except IOError as ioe:
+        # if isinstance(ioe, HttpError):
+        #     # Something went wrong server-side
+        #     print(ioe.status_code)
```
Comment (maarcingebala) [inline saleor/payment/gateways/paypal/__init__.py:112]: Same here, this part could be removed if it's not used.

File: saleor/payment/gateways/paypal/__init__.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        error_message = getattr(ioe, "status_code", repr(ioe))
+        return GatewayResponse(
+            is_success=False,
+            action_required=False,
+            amount=payment_information.amount,
+            transaction_id=payment_information.token,
+            kind=TransactionKind.CAPTURE,
```
Comment (maarcingebala) [inline saleor/payment/gateways/paypal/__init__.py:72]: Loose thought, but from what I see this implementation only supports "capture" payments without prior authorization and manual capture? Maybe we could then move this entire implementation to the `capture` method below?

File: saleor/payment/gateways/paypal/__init__.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            error=None,
+            # raw_response=response,
+        )
+
+
+def capture(payment_information: PaymentData, config: GatewayConfig) -> GatewayResponse:
+    pass
+
+
+def void(payment_information: PaymentData, config: GatewayConfig) -> GatewayResponse:
```
Comment (maarcingebala) [inline saleor/payment/gateways/paypal/__init__.py:99]: Does PayPal SDK support this operation?

**Comment by albertojacini** [comment]
I'll work on this in the weekend.

File: saleor/payment/gateways/paypal/__init__.py (hunk 1 above)
Comment (albertojacini) [reply saleor/payment/gateways/paypal/__init__.py:88]: ok

File: saleor/payment/gateways/paypal/__init__.py (hunk 5 above)
Comment (albertojacini) [reply saleor/payment/gateways/paypal/__init__.py:99]: As far as I understand, no. No void/cancel operation is supported by Paypal orders/payments Api and sdk.

**Comment by albertojacini** [comment]
Hi @maarcingebala, I pushed an updated version of the gateway. It takes in your review points and add support for partial refunds.

**Comment by Anyman552** [comment]
Hi @albertojacini, is there code for the storefront to use your feature?

**Comment by albertojacini** [comment]
Hi @Anyman552 , you can try to adapt this code: https://gist.github.com/albertojacini/c2461617418adc5f18bf07a464aa5a83 . It works in my private project.

File: saleor/payment/gateways/paypal/__init__.py (hunk 4 above)
Comment (yevheniidehtiar) [reply saleor/payment/gateways/paypal/__init__.py:72]: I am just travelling around opened PR... this one is very solid! 
@maarcingebala Ain't you worry only about logic?... i found this code could be improved a little, partially...

**Comment by fabianjost** [comment]
Hi Guys, 
@albertojacini thank you so much for the awesome work. We really need this plugin in order to use saleor.
@maarcingebala will this be merged to master any time soon?

**Comment by tlosbar** [comment]
🙏🏾 Can we hope to see this feature in the next release? This is the must have for our projects as Braintree and Stripe aren’t available for our clients

**Comment by AgentIvan** [comment]
> 🙏🏾 Can we hope to see this feature in the next release? This is the must have for our projects as Braintree and Stripe aren’t available for our clients

What about [myPOS](https://www.mypos.eu/it/faq/new-client/fees-and-limits)? I need plugin for this gateway because my Customer wants to use this for whole UE. I have started to translate myPOS [SDK](https://github.com/guug-saleor/myPOS-Checkout-SDK-Python)

**Comment by nikolas310** [comment]
@albertojacini can you give paypal storefront integration

**Comment by AgentIvan** [comment]
I think this plugin will be useful for many people

**Comment by AgentIvan** [comment]
I think this plugin will be useful for many develoers

**Comment by eddyLazar** [comment]
Hello. What is the status of this? And if any help needed I could participate. Also have problems with country support

**Comment by isik-kaplan** [comment]
Wakey wakey

**Comment by philipgierszal** [comment]
Hey guys, any updates? I would love to help if there is anything I can do :)

**Pull request closed without merging**